### PR TITLE
Fix Extensions

### PR DESCRIPTION
**Pull request by manuth**
- [x] I have read the note above about PRs contributing or fixing extensions
- [x] I have tried reaching out to the extension maintainers about publishing this extension to OpenVSX (if not, please create an issue in the extension's repo using [this template](https://github.com/open-vsx/publish-extensions/blob/HEAD/docs/external_contribution_request.md)).
- [x] This extension has an [OSI-approved OSS license](https://opensource.org/licenses) (we don't accept proprietary extensions in this repository)

## Description
As seen in the `Publish extensions to open-vsx.org` workflow, there is quite a number of extensions which is not able to build at the moment.
The goal of this PR is to fix all extensions which aren't able to build at time of writing.
All one-time fixes are guarded with version-comparisons (using the new environment variables described in the `Environment` paragraph).

Fixing all non-functioning extensions will help to detect, address and fix future issues with extensions more efficiently.
Everything that has been changed is documented below.

### Environment
In order to make applying workarounds and advanced actions in `custom`- and `prepublish`-commands easier, a few environment variables were added:

  - `EXTENSION_ID` which contains the id of the extension being published (e.g. `rebornix.ruby`)
  - `EXTENSION_PUBLISHER` the name of the publisher (e.g. `rebornix`)
  - `EXTENSION_NAME` the name of the extension (e.g. `ruby`)
  - `VERSION` the version the extension is assumed to have
  - `MS_VERSION` the latest version of the extension on microsoft's marketplace
  - `OVSX_VERSION` the latest version of the extension on Open VSX

Furthermore, a chapter about said environment variables was added to the `README`.

#### Added
  - `ms-dotnettools.vscode-dotnet-runtime` (required by some extensions which currently fail to build)

#### Fixed
Following extensions were fixed:
  - `appulate.filewatcher`  (PR https://github.com/appulate/vscode-file-watcher/pull/16, see ***Abandoned-Looking Extensions*** below)
  - `adamhartford.vscode-base64` (PR for requesting fixing changes https://github.com/adamhartford/vscode-base64/pull/8 created by @filiptronicek, see ***Abandoned-Looking Extensions*** below)
  - `arjun.swagger-viewer` (PR https://github.com/arjun-g/vs-swagger-viewer/pull/110, see ***Abandoned-Looking Extensions*** below)
  - `bierner.markdown-image-size` (PR https://github.com/mjbvz/vscode-markdown-image-size/pull/5)
  - `bmalehorn.vscode-fish` (PR https://github.com/bmalehorn/vscode-fish/pull/26)
  - `formulahendry.vscode-node-red` (PR https://github.com/formulahendry/vscode-node-red/pull/26, see ***Abandoned-Looking Extensions*** below)
  - `fraser.live-coder` (Creating PRs in GitLab doesn't seem to be possible, see ***Abandoned-Looking Extensions*** below)
  - `halcyontechltd.git-client-ibmi` (PR https://github.com/halcyon-tech/git-client-ibmi/pull/6)
  - `hashicorp.terraform`
  - `ikappas.composer` (PR https://github.com/ikappas/vscode-composer/pull/22)
  - `ikappas.phpcs` (PR https://github.com/ikappas/vscode-phpcs/pull/198)
  - `iocave.monkey-patch` (by removing the `src/test`-directory)
  - `jock.svg` (PR https://github.com/lishu/vscode-svg2/pull/107)
  - `johnsoncodehk.volar`
  - `johnsoncodehk.vscode-typescript-vue-plugin`
  - `justusadam.language-haskell`
  - `marus25.cortex-debug` (by installing the exact `package-lock.json` dependencies using `npm ci`)
  - `miguel-savignano.ruby-symbols` (PR https://github.com/MiguelSavignano/vscode-ruby-symbols/pull/13)
  - `ms-vscode.js-debug-companion` (by `git checkout`ing the (at time of writing) most recent commit)
  - `ms-vscode.mock-debug`
  - `ms-vscode.vscode-js-profile-flame`
  - `ms-vscode.vscode-js-profile-table`
  - `neikeq.godot-csharp-vscode` by performing the building in a nested docker container (PR https://github.com/godotengine/godot-csharp-vscode/pull/41)
  -  `oouo-diogo-perdigao.docthis` (PR https://github.com/oouo-diogo-perdigao/vscode-docthis/pull/10)
  - `paiqo.databricks-vscode` (PR https://github.com/paiqo/Databricks-VSCode/pull/72)
  - `peterj.proto` (PR https://github.com/peterj/vscode-protobuf/pull/2)
  - `reduckted.vscode-gitweblinks`
  - `rebornix.ruby` (PR https://github.com/rubyide/vscode-ruby/pull/810)
  - `rintoj.json-organizer` (PR https://github.com/rintoj/json-organizer/pull/2)
  - `saltud.type4py` (PR https://github.com/saltudelft/type4py-vscode-ext/pull/25)
  - `serayuzgur.crates`
  - `sodatea.velocity` (PR https://github.com/sodatea/vscode-velocity/pull/10)
  - `shuworks.vscode-table-formatter` (PR https://github.com/shuGH/vscode-table-formatter/pull/31)
  - `styled-components.vscode-styled-components`
  - `vscode-org-mode.org-mode` (PR https://github.com/vscode-org-mode/vscode-org-mode/pull/204)
  - `vshaxe.hxcpp-debugger`
  - `wmaurer.change-case` (PR https://github.com/wmaurer/vscode-change-case/pull/43)
  - `Wscats.eno` (PR https://github.com/Wscats/compile-hero/pull/151)
  - `xdebug.php-debug`
  - `xirider.livecode` (PR https://github.com/Xirider/LiveCode/pull/9)

#### Renamed
Furthermore, some extensions were renamed:
  - `andreweinand.mock-debug` replaced w/ `ms-vscode.mock-debug`
  - `felixfbecker.php-debug` replaced w/ `xdebug.php-debug`
  - `jpoissonnier.vscode-styled-components` replaced w/ `styled-components.vscode-styled-components`

### Non-Functioning Extensions
Some extensions don't work yet
Following extensions do not have an open source license assigned:
  - `ChakrounAnas.turbo-console-log` (Issue https://github.com/Chakroun-Anas/turbo-console-log/issues/160)
  - `Dannark.AndroidLauncher` (Issue https://github.com/Dannark/Android-Launcher-for-VSCode/issues/13)
  - `joelday.docthis` (Extension doesn't seem to exist anymore)
  - `ms-python.devicesimulatorexpress` (Extension has been de-listed)
  - `ms-vscode.vscode-js-profile-notebook` doesn't exist yet

### Abandoned-Looking Extensions
Some of the extensions which won't work until this PR is merged, might be abandoned.
This list contains extensions with both at least 1 Pull Request which were without any activity for at least 1 month and a repository with at least 1y of inactivity.
  - `appulate.filewatcher`
  - `adamhartford.vscode-base64` (repo inactive for about 4y)
  - `arjun.swagger-viewer` (probably inactive? repo inactive for about 9 months)
  - `formulahendry.vscode-node-red` (most likely abandoned - inactive for about 4y. Requires tampering the code)
  - `fraser.live-coder` (2y of inactivity)
  - ~~`ikappas.composer` (probably inactive? repo inactive for 8 months)~~
  - `wmaurer.change-case` (repo inactive for 6 years)

### CODE REVIEW

**Comment by filiptronicek** [comment]
@manuth thanks for taking a look at so many extensions. I see you are fixing up some things based on the version of the extensions published. Are those unmaintained extensions that would not be fixable by PRs in their respective repos?

If so, let's make PRs to encourage extension authors to make their extensions build, and if the extensions are unmaintained, I would rather publish them just once via https://github.com/open-vsx/publish-extensions/blob/master/.github/workflows/publish-once.yml. 

If you agree with publishing some extensions once, could you make a list somewhere 🙏?

**Comment by manuth** [comment]
Thahks for having a look at this PR 😄 
I added the version guard to make sure the workaround won't be applied on upcoming versions. That way, the `custom` property won't destroy (or tamper) upcoming versions.

I'll make sure to add a list with abandoned extensions and build commands to have them published once.

**Comment by filiptronicek** [comment]
Oh wow, this is huge 😄. Thanks so much for taking the time, @manuth. If that is ok with you, I will not be merging the PR right away. I will try to fix as many extensions as possible that I can do one-time, because that is going to help a lot with the repository cleanup.

I see that you have also submitted a PR for fixing https://github.com/ikappas/vscode-phpcs. I have been trying to package it for a couple of months now, but even in the PR you submitted (https://github.com/ikappas/vscode-phpcs/pull/198) I could not get it to work. Can you share more details on packaging this one? The process I did is as follows:
```bash
npm i -g rimraf npm-run-all
npm i
npm run compile
```

From that I got 
```
> phpcs-server@1.0.5 installServer
> installServerIntoExtension ../phpcs ./package.json ./tsconfig.json

sh: 1: installServerIntoExtension: not found
ERROR: "compile:server" exited with 127.
```

Running the compile command in the `phpcs/` subdirectory sadly only showed me some `tsc` errors. It is one of the biggest extensions that we still don't have on OpenVSX, and since it's unmaintained it could be a quick win, theoretically 😄.

> ikappas.phpcs (PR https://github.com/ikappas/vscode-phpcs/pull/198, might be a good idea to swap to shevaua.phpcs?)

Maybe, but if no further development occurs, removing it is going to be enough in my opinion (the fork is only attempting to fix the upstream, with no other changes).

**Comment by manuth** [comment]
Thanks a lot for the appreciation 😄 
I had a look through it and it looks like from `npm@6` (the one currently used in this repo's `publish extensions` workflow vs `npm@8.5.5` (the most recent version) changed slightly.

While in `npm@6` `preinstall` was executed BEFORE installing the dependencies to `node_modules`, in `npm@8` (whyever) `preinstall` is executed AFTER installing the dependencies to `node_modules`.

Because of that, the modules are basically being installed and deleted again right afterwards (cause of the `rimraf node_modules` call).
Therefore, all the commands from installed dependencies (such as `installServerIntoExtension` from dependency `vscode-languageserver`) won't work.

I removed the `preinstall` scripts and found myself able to build the extension with both `npm@6` and `npm@8`.
I adjusted the PR accordingly.

**Comment by filiptronicek** [comment]
> I removed the preinstall scripts and found myself able to build the extension with both npm@6 and npm@8.
I adjusted the PR accordingly.

Great stuff! I tried and can produce a `.vsix` and it worked 🎉. I am a bit worried about publishing it because they vary quite a bit in size (the smaller one is the one I managed to produce):
```
gitpod /workspace/vscode-phpcs/phpcs (patch-1) $ du -sh phpcs-1.0.5.vsix ikappas.phpcs-1.0.5\ \(1\).vsix 
1.2M    phpcs-1.0.5.vsix
5.0M    ikappas.phpcs-1.0.5 (1).vsix
```

I guess we should try the extension by testing, I'll try it out. So far, I have gotten errors about PHP CS missing as a global dependency.

**Comment by manuth** [comment]
@filiptronicek I just checked it out and found what's causing the huge difference in size.
The `phpcs` extension (probably by mistake) has `vscode` in its `dependencies` rather than `devDependencies` causing a huge load of files in the `.vsix`s `node_modules` directory.

So I think we can safely ignore the difference.

Thanks for taking care 😄 
Also I'm through with all non-functioning extensions very soon (I'm only 5 short)

**Comment by filiptronicek** [comment]
Awesome-sauce! I'll go ahead and publish phpcs right away. it's crazy to think about how long we wanted to do this, so thanks a lot @manuth!

**Comment by manuth** [comment]
@filiptronicek I'm finished now apart from the Godot C# extension. I'll finish up the description tomorrow and - if I'm not able to get Godot C# to work - remove Godot C#.

Thanks for your patience and thanks for standing with me :smile:

**Comment by manuth** [comment]
@filiptronicek I'm finished with all extensions now.
I even found myself able to get `neikeq.godot-csharp-vscode` to build.

Hope you'll find some time to have a look at it and merge it soon 😄 
I made sure that all workaround commands have a `|| true` guard at the end so we should be fine!

With your `custom` feature added, I'm looking forward to see all them great extensions (especially Godot for C#, lol) in the Open VSX marketplace!

Thanks for the effort 😄

File: extensions.json [hunk 1]
```diff
@@ -68,15 +83,30 @@
...
     "repository": "https://github.com/yarnpkg/berry",
     "location": "packages/vscode-zipfs",
     "prepublish": "sed -ri \"s_ssh://git@github.com/yarnpkg/berry.git_https://github.com/yarnpkg/berry_g\" packages/vscode-zipfs/package.json"
   },
   "arjun.swagger-viewer": {
-    "repository": "https://github.com/arjun-g/vs-swagger-viewer"
+    "repository": "https://github.com/arjun-g/vs-swagger-viewer",
+    "location": ".",
+    "custom": [
+      "[ $VERSION = '3.1.2' ] && sed -i '/\"@types\\/socket\\.io\": \".*\"/d' package.json || true",
```
Comment (filiptronicek) [inline extensions.json:98]: I'm thinking about whether it is necessary here, because `sed` could theoretically replace these values on any version of the extension, and if it is fixed, sed would just continue without replacing, correct?

I am just a bit worried about a lot of maintenance after these extensions get updates and a lot of cleanups necessary afterwards. If we only want to do some fixes for one version only, we could run the publish-extensions workflow on this branch one-time so that those versions are all published, making the fixes being in `master` unnecesary.

Comment (manuth) [reply extensions.json:98]: That is correct. A command which is prefixed with such a version guard will only execute if the extension has a specific version.
But I think running the publish task and cleaning up the `extensions.json` file before merging sounds like a great idea 😄 

Are there any version-guarded commands that should remain in the `extension.json` file?
I think keeping the one of `serayuzgur.crates` inside might make sense as the `package.json` file in this repository isn't properly versioned and there are no git tags either causing the publish-workflow to `git checkout` any random tag which is contained in the `msVersion` name (by coincidence)

**Comment by filiptronicek** [comment]
I published some one-time extensions (I'll keep the list updated):
- `formulahendry.vscode-node-red`
- `adamhartford.vscode-base64`

**Comment by filiptronicek** [comment]
I took the liberty to remove the extensions I've published

**Comment by manuth** [comment]
Well - let's just hope they won't push another version, then 😅
Thanks for the effort, that's awesome 😄

**Comment by filiptronicek** [comment]
> Well - let's just hope they won't push another version, then 😅

I will try to let the authors know that we removed it from here, because it might create confusion with people not getting why their extension is on OpenVSX and is not updating.

File: extensions.json [hunk 2]
```diff
@@ -430,7 +422,9 @@
     "repository": "https://github.com/halcyon-tech/code-for-ibmi"
   },
   "halcyontechltd.git-client-ibmi": {
-    "repository": "https://github.com/halcyon-tech/git-client-ibmi"
+    "repository": "https://github.com/halcyon-tech/git-client-ibmi",
+    "location": ".",
+    "prepublish": "[ $VERSION = '0.1.0' ] && mv LICENCE LICENSE || true"
```
Comment (filiptronicek) [inline extensions.json:427]: It seems this change does not actually do anything, so I think we can remove it

```suggestion
    "repository": "https://github.com/halcyon-tech/git-client-ibmi",
```

File: extensions.json [hunk 3]
```diff
@@ -1042,7 +1137,13 @@
     "repository": "https://github.com/shrey150/vscode-javafx-support"
   },
   "shuworks.vscode-table-formatter": {
-    "repository": "https://github.com/shuGH/vscode-table-formatter"
+    "repository": "https://github.com/shuGH/vscode-table-formatter",
+    "location": ".",
```
Comment (filiptronicek) [inline extensions.json:1141]: Just FYI @manuth all extensions have the location as `.` by default. Let me know if you had any problems with not specifying this property.

Comment (manuth) [reply extensions.json:1141]: I got an error about `path` expecting a `string` but receiving `null`

I assume the problem's source is here:
https://github.com/open-vsx/publish-extensions/blob/297df0a13c62eb9a0f0d0fe037430810a99476c7/publish-extension.js#L47

File: extensions.json (hunk 2 above)
Comment (manuth) [reply extensions.json:427]: It does - previously, this extension wasn't published because ovsx couldn't find an open source license.
I assume renaming the file properly will help, does it?

File: extensions.json (hunk 3 above)
Comment (filiptronicek) [reply extensions.json:1141]: Oh, right, let me fix that real quick 😄

Comment (manuth) [reply extensions.json]: Done! 😄

File: extensions.json [hunk 4]
```diff
@@ -52,10 +49,11 @@
...
-    "repository": "https://github.com/eamodio/vscode-tsl-problem-matcher"
-  },
-  "andreweinand.mock-debug": {
-    "repository": "https://github.com/microsoft/vscode-mock-debug"
+    "repository": "https://github.com/eamodio/vscode-tsl-problem-matcher",
+    "custom": [
+      "yarn install",
+      "vsce package -o extension.vsix"
+    ]
   },
```
Comment (filiptronicek) [inline extensions.json:57]: I just checked this particular extension, and it seems it's been resulting in an `Internal Server Error` for close to a year now. I will check with the team internally at Eclipse to see what's up. The extension is resolving just fine apparently so I don't think any custom config will be needed.

File: extensions.json [hunk 5]
```diff
@@ -430,7 +420,8 @@
     "repository": "https://github.com/halcyon-tech/code-for-ibmi"
   },
   "halcyontechltd.git-client-ibmi": {
-    "repository": "https://github.com/halcyon-tech/git-client-ibmi"
+    "repository": "https://github.com/halcyon-tech/git-client-ibmi",
+    "prepublish": "[ $VERSION = '0.1.0' ] && mv LICENCE LICENSE || true"
```
Comment (filiptronicek) [inline extensions.json:424]: Does `mv LICENSE LICENSE` do something here?

File: extensions.json (hunk 4 above)
Comment (manuth) [reply extensions.json:57]: Okay, I'll remove this section then

File: extensions.json (hunk 5 above)
Comment (manuth) [reply extensions.json:424]: Yes - the file is incorrectly called `LICENCE`, thus `ovsx` is unable to detect the license.
Renaming the file to `LICENSE` fixes the issue.

Comment (filiptronicek) [reply extensions.json:420]: Looks like the author has setup OpenVSX publishing, and it just failing because of the license: https://github.com/halcyon-tech/git-client-ibmi/blob/master/.github/workflows/ci.yaml.

File: extensions.json [hunk 6]
```diff
@@ -981,7 +1044,19 @@
     "repository": "https://github.com/richie5um/vscode-sort-json"
   },
   "rintoj.json-organizer": {
```
Comment (filiptronicek) [inline extensions.json:1036]: One-time published this, thanks for making it easy to package up 😄!

File: extensions.json [hunk 7]
```diff
@@ -999,7 +1049,8 @@
     "repository": "https://github.com/ryu1kn/vscode-edit-with-shell"
   },
   "saltud.type4py": {
-    "repository": "https://github.com/saltudelft/type4py-vscode-ext"
+    "repository": "https://github.com/saltudelft/type4py-vscode-ext",
+    "prepublish": "[ $VERSION = '0.2.0' ] && npm run package-web || true"
```
Comment (filiptronicek) [inline extensions.json:1053]: I think we can revert the changes - we can just have it fail until the next version, so that we don't have to remove the version check afterwards.

Comment (manuth) [reply extensions.json:1053]: I think it might be a good idea to perform an individual publish, then
What do you think?

Comment (filiptronicek) [reply extensions.json:1053]: Sounds good, will do that.

Comment (manuth) [reply extensions.json:1053]: Awesome! Thanks for your effort 😄

Comment (filiptronicek) [reply extensions.json:1053]: https://github.com/open-vsx/publish-extensions/runs/5821413325?check_suite_focus=true#step:5:4 - published 0.2.0

**Comment by manuth** [comment]
@filiptronicek is there something more to be done or to consider before we can get this PR merged?
I'm ready from my side 😄

**Comment by filiptronicek** [comment]
@manuth I am going to take another look now, I'd love to get this merged soon, just have some other things to finish up 👀.

**Comment by manuth** [comment]
> @manuth I am going to take another look now, I'd love to get this merged soon, just have some other things to finish up 👀.

No worries I just wanted to make sure I'm not missing something 😄 
Thank you so much for taking care ^^

**Review by filiptronicek — APPROVED** [review]
Reviewed the rest and made some changes. This PR LGTM! So glad you made the time for it, thanks again @manuth.

**Comment by filiptronicek** [comment]
I have to sadly remove the very-hard-to-package-up the `neikeq.godot-csharp-vscode` extension, because it depends on a proprietary extension from Microsoft, `ms-dotnettools.csharp` [[1](https://github.com/godotengine/godot-csharp-vscode/blob/0f273c273ee5045ae1ddbca9dd8dd64a7a73adb7/package.json#L48)]. This makes the publishing fail [[1](https://github.com/open-vsx/publish-extensions/runs/6174399836?check_suite_focus=true#step:8:6805)] and sadly there is so far no good way to get around it AFAIK :/ 

cc @manuth

**Comment by manuth** [comment]
Before you do, I'll go try whether the Open Source alternative of the OmniSharp extension.

Maybe it works with it out of the box. I really hope it does 🥲😂

**Comment by filiptronicek** [comment]
Let's see @manuth, as I already commited the removal, should I revert it?

**Comment by manuth** [comment]
@filiptronicek Thanks for taking care 😄 
I managed it to get it to work by replacing the extension dependencies with `muhammad-sammy.csharp`.
Funny enough, I even got the "C# Tools for Godot" debugging to run without `muhammad-sammy.csharp`, so I'm not sure what it's in the dependencies for (perhaps for a better C# coding experiences).

Could you please revert the changes?
I'll open up a PR adding a workaround which replaces the extension dependencies with `muhammad-sammy.csharp` asap!

Thanks a ton 😄

**Comment by filiptronicek** [comment]
@manuth done :) https://github.com/open-vsx/publish-extensions/commit/d0889ed64140b72754cb0bd9d93f498c9b5cd493
